sql/query/expr_test.go

```diff
@@ -0,0 +1,189 @@
+// Copyright 2015 The Cockroach Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the License. See the AUTHORS file
+// for names of contributors.
+//
+// Author: Peter Mattis (peter@cockroachlabs.com)
+
+package query
+
+import (
+	"testing"
+
+	"github.com/cockroachdb/cockroach/sql/parser"
+	"github.com/cockroachdb/cockroach/sql/sqlwire"
+	"github.com/cockroachdb/cockroach/testutils"
+)
+
+func dInt(v int64) sqlwire.Datum {
+	return sqlwire.Datum{IntVal: &v}
+}
+
+func dUint(v uint64) sqlwire.Datum {
+	return sqlwire.Datum{UintVal: &v}
+}
+
+func dFloat(v float64) sqlwire.Datum {
+	return sqlwire.Datum{FloatVal: &v}
+}
+
+func dBytes(v []byte) sqlwire.Datum {
+	return sqlwire.Datum{BytesVal: v}
+}
+
+func dString(v string) sqlwire.Datum {
+	return sqlwire.Datum{StringVal: &v}
+}
+
+func TestEvalExpr(t *testing.T) {
+	testData := []struct {
+		expr     string
+		expected string
+		env      Env
+	}{
+		// Bitwise operators.
+		{`1 & 3`, `1`, nil},
+		{`1 | 3`, `3`, nil},
+		{`1 ^ 3`, `2`, nil},
+		// Bitwise operators convert their arguments to ints.
```

Comment:
Personally I'd rather have this be an error; we should avoid implicit conversions that lose data.